content/academy/switching_to_typescript/unknown_and_type_casting.md

```diff
@@ -0,0 +1,103 @@
+
+![Replacing "any" with "unknown" from the above snippet]({{@asset switching_to_typescript/images/replace-with-unknown.webp}})
+
+Even this will result in the same error:
+
+```TypeScript
+// This results in a compiler error!
+let userInput: unknown;
+let savedInput: string;
+
+userInput = 'hello world!';
+
+savedInput = userInput;
+```
+
+## [](#type-guards) Type guards
+
+In order to make the code above not throw any compiler errors, we can use a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html), which is just a check that happens at runtime to ensure that the type is in fact what it should be.
+
+```TypeScript
+let userInput: unknown;
+let savedInput: string;
+
+userInput = 5;
+
+// This if statement is called a "type guard"
+// No more error! TypeScript is smart enough to understand
+// what this if statement is doing, and removes the error
+if (typeof userInput === 'string') {
+    savedInput = userInput;
+}
+```
+
+This works, and in fact, it's the most optimal solution for this use case. But what if we were 100% sure that the value stored in `userInput` was a string? Thats when **type casting** comes in handy.
+
+## [](#type-casting) Type casting
+
+Despite the fancy name, [type casting](https://www.typescripttutorial.net/typescript-tutorial/type-casting/) is a simple concept based around a single keyword: `as`. We usually use this on values that we can't control the return type of, or values that we're sure have a certain type, but TypeScript needs a bit of help understanding that.
+
+```TypeScript
+let userInput: unknown;
+let savedInput: string;
+
+userInput = 'hello world!';
+
+// No more error! We've told TypeScript to treat "userInput"
+// as a string, despite the fact that its original type is
+// the "unknown" type
+savedInput = userInput as string;
+```
+
+## [](#final-thoughts) Final thoughts
+
+Even though you've learned about them in the same lesson, type guards and type casting are inherently very different things with different use cases.
+
+**Type guards** make a runtime check of whether or not a value passes a check that determines that it can be safely used as a certain type. They are great when dealing with values that might hold inconsistent data types (such as user input) where you aren't 100% sure if a certain property will exist.
+
+**Type casting** tells TypeScript to take a value of one type and to treat it as if it were another type. No runtime checks are made. A common use case is casting the response body of an API call (usually has the `any` type depending on what you're using to fetch the data) to a custom type to receive TypeScript support on the data.
+
```

Comment:
Would mention here also the `!` token which lets you force the type if you know it is there but the compiler does not.

Reply:
Worth noting that the term "type casting" should be in fact ["type assertions"](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#type-assertions), there is no casting happening, and it does not work with "values" at all. We all use the term "casting" as it feels familiar (and I would indeed mention it here too, but not as a headline), but it does not cast any value, it changes the type. Let's use the right terms so its easier to find appropriate docs for it.

And agreed we should mention the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) too.

Reply:
both resolved